### PR TITLE
Support stop and shutdown in flow Control

### LINKED ISSUE #135: Now way to perform correct shutdown
There is no way to perform correct `Consumer` shutdown in the current API. 

Pseudocode to illustrate:

```scala
val control = Consumer.commitableSource(config)
  .mapAsync(10)(longAsyncTaskWhichTakes1SecondOrSo)
  .mapAsync(batchCommit) //commit once per second
  .to(Sink.ignore)
  .run()
```

The code above starts stream processing. Now there is only `control.stop()` to shut down the `Consumer`. This method waits for no commits in progress and shut down kafka consumer regardless messages processed by my async task.

With this semantic there is no way to perform correct shutdown even when everything is ok.

One of solution can be introducing `stop(finiteDuration)` method, which completes the `Source` immediately, but keeps kafka consumer online for given duration.

### PR DESCRIPTION
**Pull request by 13h3r**
Introducing `stop` and `shutdown` in `Control` to be able to perform correct shutdown. Fixes #135

### CODE REVIEW

File: core/src/main/scala/akka/kafka/scaladsl/Consumer.scala [hunk 1]
```diff
@@ -111,18 +111,27 @@ object Consumer {
...
    */
   trait Control {
+
+    /**
+     * Stop producing messages from the `Source`. This does not stop underlying kafka consumer
+     * and does not unsubscribe from any topics/partitions.
+     *
+     * Call [[#shutdown]] to close consumer
+     */
+    def stop(): Unit
```
Comment (kciesielski) [inline core/src/main/scala/akka/kafka/scaladsl/Consumer.scala]: Shouldn't `stop` return a `Future` too?

Comment (13h3r) [inline core/src/main/scala/akka/kafka/scaladsl/Consumer.scala]: If it complete when stage is shutdown then it never completes until someone call `shutdown`. This is look confusing for me. 

Typical case for me is:
- call `stop`
- wait for correct stream completion
- call shutdown at stream completion or after some timeout for stream completion.

Comment (patriknw) [inline core/src/main/scala/akka/kafka/scaladsl/Consumer.scala]: The stop itself (stop reading) is an async operation so I think it makes sense to return `Future[Done]` to signal that the stop operation has been done, i.e. no more messages will be emitted. Might be needed when chaining things.

Comment (13h3r) [inline core/src/main/scala/akka/kafka/scaladsl/Consumer.scala]: Guys, agreed with you. Will implementit with Patrik's semantic

**Comment by 13h3r** [comment]
@patriknw @kciesielski added `Future` to `stop`. Please, take a look

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 2]
```diff
@@ -435,9 +436,11 @@ private[kafka] abstract class ConsumerStageLogic[K, V, Out](
   // impl of Consumer.Control that is called from outside via materialized value
   override def isShutdown: Future[Done] = shutdownPromise.future
 
-  override def stop(): Unit = {
+  override def stop(): Future[Done] = {
     log.debug("Stop producing messages")
-    stopCallback.invoke(())
+    val result = Promise[Done]
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: There is a problem with this. If the stage is already shutdown, this promise will never be completed. Better to implement it as is done for `shutdown`, i.e. with a `val stopPromise` in the stage logic.

Comment (13h3r) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: `result` will be completed in callback so every promise will be completed, but I agreed that calling `complete` multiple times smells.

Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: nope, the callback will not executed if the stage is stopped

Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: and the I mean as in `postStop`

Comment (13h3r) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: Got it. Nice catch!

**Comment by 13h3r** [comment]
Any comments on this?

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 3]
```diff
@@ -314,19 +314,18 @@ private[kafka] abstract class ConsumerStageLogic[K, V, Out](
...
   private var pollScheduled = false
-  private var stopped = false
 
   private val shutdownCallback = getAsyncCallback[Unit] { _ => shutdownInternal() }
-  private val stopCallback = getAsyncCallback[Promise[Done]] { p =>
+  private val stopCallback = getAsyncCallback[Unit] { p =>
     complete(out)
-    stopped = true
-    p.success(Done)
+    stopPromise.success(Done)
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: should be `trySuccess`

**Comment by patriknw** [comment]
LGTM, after that `trySuccess`, please squash before merging

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 4]
```diff
@@ -419,22 +418,27 @@ private[kafka] abstract class ConsumerStageLogic[K, V, Out](
         None
     }
     // close() is blocking (stupid api that doesn't take timeout parameters)
-    // WakeupException will be thrown by close() if the wakeup from stopTimeout is triggered
+    // WakeupException will be thrown by close() if the wakeup from shutdownTimeout is triggered
     Try(consumer.close())
-    stopTimeoutTask.foreach(_.cancel())
-    stoppedPromise.success(Done)
+    shutdownTimeoutTask.foreach(_.cancel())
+    shutdownPromise.success(Done)
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: @13h3r `stopPromise.trySuccess(Done)` here also

**Comment by 13h3r** [comment]
ping @patriknw

File: core/src/main/scala/akka/kafka/internal/ConsumerStage.scala [hunk 5]
```diff
@@ -419,22 +418,27 @@ private[kafka] abstract class ConsumerStageLogic[K, V, Out](
...
     }
     // close() is blocking (stupid api that doesn't take timeout parameters)
-    // WakeupException will be thrown by close() if the wakeup from stopTimeout is triggered
+    // WakeupException will be thrown by close() if the wakeup from shutdownTimeout is triggered
     Try(consumer.close())
-    stopTimeoutTask.foreach(_.cancel())
-    stoppedPromise.success(Done)
+    shutdownTimeoutTask.foreach(_.cancel())
+    shutdownPromise.trySuccess(Done)
     super.postStop()
```
Comment (patriknw) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: @13h3r perhaps you missed my previous comment. There is a need for `stopPromise.trySuccess(Done)` here also.

Comment (13h3r) [inline core/src/main/scala/akka/kafka/internal/ConsumerStage.scala]: Miss it, sorry. Fixed now

**Comment by patriknw** [comment]
thanks

**Pull request merged**